src/Processors/Formats/Impl/ORCBlockInputFormat.cpp

```diff
@@ -39,23 +38,53 @@ Chunk ORCBlockInputFormat::generate()
     if (!file_reader)
         prepareReader();
 
-    if (stripe_current >= stripe_total)
-        return res;
+    while (true)
+    {
+        if (!batch_reader)
+        {
+            arrow::Status reader_status = file_reader->NextStripeReader(
+                DBMS_DEFAULT_BUFFER_SIZE, include_indices, &batch_reader);
+            if (!reader_status.ok())
+                throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA,
+                                       "Failed to create batch reader: {}",
+                                       reader_status.ToString());
+        }
 
-    std::shared_ptr<arrow::RecordBatch> batch_result;
-    arrow::Status batch_status = file_reader->ReadStripe(stripe_current, include_indices, &batch_result);
-    if (!batch_status.ok())
-        throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA,
-                               "Error while reading batch of ORC data: {}", batch_status.ToString());
+        if (!batch_reader)
+            break;
 
-    auto table_result = arrow::Table::FromRecordBatches({batch_result});
-    if (!table_result.ok())
-        throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA,
-                               "Error while reading batch of ORC data: {}", table_result.status().ToString());
+        std::shared_ptr<arrow::RecordBatch> batch_result;
+        arrow::Status batch_status = batch_reader->ReadNext(&batch_result);
+        if (!batch_status.ok())
+            throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA,
+                                   "Error while reading batch of ORC data: {}",
+                                   batch_status.ToString());
+
+        if (!batch_result || !batch_result->num_rows())
+        {
+            batch_reader.reset();
+            continue;
+        }
 
-    ++stripe_current;
+        auto table_result = arrow::Table::FromRecordBatches({batch_result});
+        if (!table_result.ok())
+            throw ParsingException(ErrorCodes::CANNOT_READ_ALL_DATA,
+                                   "Error while reading batch of ORC data: {}",
+                                   table_result.status().ToString());
+
+        using NameToColumnPtr = std::unordered_map<std::string, std::shared_ptr<arrow::ChunkedArray>>;
+
+        NameToColumnPtr name_to_column_ptr;
+        for (const auto& column_name : column_names)
+        {
+            arrow::ArrayVector vec = {batch_result->GetColumnByName(column_name)};
+            std::shared_ptr<arrow::ChunkedArray> arrow_column = std::make_shared<arrow::ChunkedArray>(vec);
+            name_to_column_ptr[column_name] = arrow_column;
+        }
+        arrow_column_to_ch_column->arrowColumnsToCHChunk(res, name_to_column_ptr);
+        break;
```

Comment:
What is the reason of not using `arrowTableToCHChunk` as before? Now we create arrow table `table_result`, don't use it at all and create `name_to_column_ptr` from `batch_result`.

Reply:
My intention was to remove arrow::Table::FromRecordBatches at all because it doubles number of going throw rows, but after rewriting this part of code several times seems like I left arrow::Table::FromRecordBatches unremoved. Need to delete rows from 70-75.